paprika/tleap.py

```diff
@@ -0,0 +1,574 @@
+        # smaller, by an increasingly smaller amount, (if we've taken more than one step since the last exponent change)
+        if self.wat_added_history[-2] < self.target_waters and self.wat_added_history[-1] > self.target_waters and self.cyc_since_last_exp_change > 1:
+            log.debug('Adjustment loop 1')
+            self.exponent -= 1
+            self.cyc_since_last_exp_change = 0
+            self.buffer_value = self.buffer_val_history[-1] + -5 * (10**self.exponent)
+        # If the number of waters was greater than the target and is now less than the target, make the buffer a bit
+        # bigger, by an increasingly smaller amount, (if we've taken more than one step since the last exponent change)
+        elif self.wat_added_history[-2] > self.target_waters and self.wat_added_history[-1] < self.target_waters and self.cyc_since_last_exp_change > 1:
+            log.debug('Adjustment loop 2')
+            self.exponent -= 1
+            self.cyc_since_last_exp_change = 0
+            self.buffer_value = self.buffer_val_history[-1] + 5 * (10**self.exponent)
+        # If the last two rounds of solvation have too many waters, make the buffer smaller...
+        elif self.wat_added_history[-2] > self.target_waters and self.wat_added_history[-1] > self.target_waters:
+            log.debug('Adjustment loop 3')
+            self.buffer_value = self.buffer_val_history[-1] + -1 * (10**self.exponent)
+            self.cyc_since_last_exp_change += 1
+        # If the number of waters was greater than the target and is now less than the target, make the buffer a bit
+        # bigger ...
+        elif self.wat_added_history[-2] > self.target_waters and self.wat_added_history[-1] < self.target_waters:
+            log.debug('Adjustment loop 4')
+            self.buffer_value = self.buffer_val_history[-1] + 1 * (10**self.exponent)
+            self.cyc_since_last_exp_change += 1
+        # If the number of waters was less than the target and is now greater than the target, make the buffer a bit
+        # smaller ...
+        elif self.wat_added_history[-2] < self.target_waters and self.wat_added_history[-1] > self.target_waters:
+            log.debug('Adjustment loop 5')
+            self.buffer_value = self.buffer_val_history[-1] + -1 * (10**self.exponent)
+            self.cyc_since_last_exp_change += 1
+        # If the last two rounds of solvation had too few waters, make the buffer bigger...
+        elif self.wat_added_history[-2] < self.target_waters and self.wat_added_history[-1] < self.target_waters:
+            log.debug('Adjustment loop 6')
+            self.buffer_value = self.buffer_val_history[-1] + 1 * (10**self.exponent)
+            self.cyc_since_last_exp_change += 1
+        else:
+            raise Exception(
+                "The buffer_values search died due to an unanticipated set of variable values"
+            )
```

Comment:
Six loops seems complicated, but if it works efficiently, then LGTM.

Reply:
yup, the pull requrest travis failure in test_build is evidence of why we need to keep the logic in some format.